Cleanup in api Ruby: Use Dir.chdir context guard, common.run_inline. Use spawn() for all of Workbench.

markfickett opened the pull request.

danqrodney commented on api/libproject/devstart.rb:
```diff
@@ -205,36 +211,29 @@ def do_run_migrations(creds_file, project)
-    unless system("cat db/create_db.sql | envsubst > #{create_db_file.path}")
-      raise("Error generating create_db file; exiting.")
-    end
+    common.run_inline("cat db/create_db.sql | envsubst > #{create_db_file.path}")
```
My experience was that commands like this (where we're actually executing multiple commands in the shell) don't work with run_inline. That's why I was using system in places.

markfickett replied:
OK, I reverted the ones with pipes/redirects to use system() and pulled out a helper for that case to reduce unless..raise duplication, echo the commands as they're run, and have a single place to comment on why system() is required.

I also elided some of the temp files, since I think they can just be piped also -- any reason not to do that?

danqrodney commented on api/libproject/devstart.rb:
```diff
@@ -205,36 +211,29 @@ def do_run_migrations(creds_file, project)
+    common.run_inline("cat db/create_db.sql | envsubst > #{create_db_file.path}")
     puts "Creating database if it does not exist..."
-    unless system("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
+    common.run_inline("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
```
I'm not sure that < and > work with run_inline, either.

danqrodney commented on api/libproject/devstart.rb:
```diff
@@ -205,36 +211,29 @@ def do_run_migrations(creds_file, project)
-    unless system("cat db/drop_db.sql | envsubst > #{drop_db_file.path}")
-      raise("Error generating drop_db file; exiting.")
-    end
+    common.run_inline("cat db/drop_db.sql | envsubst > #{drop_db_file.path}")
```
See above.

danqrodney commented on api/libproject/devstart.rb:
```diff
@@ -205,36 +211,29 @@ def do_run_migrations(creds_file, project)
+    common.run_inline("cat db/drop_db.sql | envsubst > #{drop_db_file.path}")
     puts "Dropping database..."
-    unless system("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
+    common.run_inline("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
```
See above.

markfickett commented:
Dan, I assumed the reason you stuck with system() was because of the shell
v. no-shell behavior I commented on at the top of the file. Were you using
the %W{..} style argument, passing as an array? If you used the array-style
argument, system() doesn't run the command in a shell, which also broke for
me in a variety of places. However, passing a string, run_inline behaves
like system().

I tested with connect_to_db, which does work with the change to
run_inline(".."); I avoided running docker_clean because last time I had to
restart my machine after running that; and I didn't run drop_cloud_db
because I assume that would cause some problems and there isn't a dry-run
mode.

On Thu, Oct 12, 2017 at 4:14 PM, danqrodney <notifications@github.com>
wrote:

> *@danqrodney* commented on this pull request.
>
> Have you tested all of these commands?
>
> All the places where I didn't use run_inline in the past were for a
> particular reason... I wouldn't change any of them without testing.
>
> The Dir.chdir() change seems fine.
> ------------------------------
>
> In api/libproject/devstart.rb
> <https://github.com/all-of-us/workbench/pull/184#discussion_r144397444>:
>
> > @@ -205,36 +211,29 @@ def do_run_migrations(creds_file, project)
>    common = Common.new
>    create_db_file = Tempfile.new("#{project}-create-db.sql")
>    begin
> -    unless system("cat db/create_db.sql | envsubst > #{create_db_file.path}")
> -      raise("Error generating create_db file; exiting.")
> -    end
> +    common.run_inline("cat db/create_db.sql | envsubst > #{create_db_file.path}")
>
> My experience was that commands like this (where we're actually executing
> multiple commands in the shell) don't work with run_inline. That's why I
> was using system in places.
> ------------------------------
>
> In api/libproject/devstart.rb
> <https://github.com/all-of-us/workbench/pull/184#discussion_r144397496>:
>
> >      puts "Creating database if it does not exist..."
> -    unless system("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
> +    common.run_inline("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
>
> I'm not sure that < and > work with run_inline, either.
> ------------------------------
>
> In api/libproject/devstart.rb
> <https://github.com/all-of-us/workbench/pull/184#discussion_r144397602>:
>
> >    begin
> -    unless system("cat db/drop_db.sql | envsubst > #{drop_db_file.path}")
> -      raise("Error generating drop_db file; exiting.")
> -    end
> +    common.run_inline("cat db/drop_db.sql | envsubst > #{drop_db_file.path}")
>
> See above.
> ------------------------------
>
> In api/libproject/devstart.rb
> <https://github.com/all-of-us/workbench/pull/184#discussion_r144397616>:
>
> >      puts "Dropping database..."
> -    unless system("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
> +    common.run_inline("mysql -u \"root\" -p\"#{ENV["MYSQL_ROOT_PASSWORD"]}\" --host 127.0.0.1 "\
>
> See above.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/all-of-us/workbench/pull/184#pullrequestreview-69050253>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASy7s4aU0e_sU63pu2uxefbo4VYRs3sks5srnMzgaJpZM4P3jPt>
> .
>

danqrodney commented:
Note: I had requested from @dmohs the ability to turn on / off printing of commands to the console with a debug flag in the common.run_inline calls. (It is useful to be able to view commands for debugging purposes; it just shouldn't be the default.) But that hasn't been implemented yet AFAIK.

markfickett commented:
I want to see the commands echoed. That's part of why I was making the switch, since I was trying to debug stuff as I refactored.

Changed to redact passwords. We have the power to edit any code! Bwahahahahahahahaha. (@dmohs, what do you think of this approach? I can send you a PR for the common lib if it looks good.)

danqrodney commented:
OK. I personally don't like seeing commands when I'm not debugging; it's noisy and obscures actual errors. So if you're going to do this, I'd prefer a flag which can be set indicating whether debugging is on or off (defaulting to off.)

Having passwords hidden when we are displaying commands for debugging makes sense.

markfickett commented:
Thanks!

Good news: `project.rb docker-clean` works fine. Bad news: Have to reboot my machine again. Adding a warning to that command.

danqrodney commented on api/libproject/devstart.rb:
```diff
@@ -468,7 +468,8 @@ def get_test_service_account_creds(*args)
     "Removes docker containers and volumes, allowing the next `dev-up` to" \
     " start from scratch (e.g., the database will be re-created). Includes ALL" \
-    " docker images, not just for the API.",
+    " docker images, not just for the API. Warning: You will probably have to" \
```
Why would you need to reboot? 

I know you ran into that problem once, but I have run this target a bunch of times and never had that issue...

danqrodney replied:
Just saw your comment. Weird.

markfickett replied:
Maybe it's `Mark will probably` and not generic `You`. I'm getting the same error as before.

dmohs commented on api/libproject/devstart.rb:
```diff
@@ -10,9 +14,11 @@
+def create_parser(command_name)
   OptionParser.new do |parser|
-    parser.banner = "Usage: ./project.rb #{command} [options]"
+    parser.banner = "Usage: ./project.rb #{command_name} [options]"
```
Good rename.

dmohs commented on api/libproject/devstart.rb:
```diff
@@ -200,52 +206,53 @@ def run_cloud_sql_proxy(project, creds_file)
+  common = Common.new
+  command_to_echo = command_string.clone
+  if to_redact
+    command_to_echo.sub! to_redact, "*" * to_redact.length
```
This is cool!

dmohs commented on api/libproject/devstart.rb:
```diff
@@ -200,52 +206,53 @@ def run_cloud_sql_proxy(project, creds_file)
   return pid
 end
 
+# Common.run_inline uses spawn() which doesn't handle pipes/redirects.
```
Note that there is `common.pipe` for this purpose: https://github.com/dmohs/project-management/blob/master/common.rb#L144
